### PR TITLE
[TypeSpec Requirement] Fail if spec contains OpenAPI generated from TypeSpec, but no tspconfig.yaml

### LINKED ISSUE #27268: [TypeSpec Requirement] Fail if spec contains OpenAPI generated from TypeSpec, but no tspconfig.yaml
## Examples
* https://github.com/Azure/azure-rest-api-specs/pull/28090

### PR DESCRIPTION
**Pull request by mikeharder**
- Fixes #27268

Validation build: https://github.com/Azure/azure-rest-api-specs/pull/28444/checks?check_run_id=23085385602

image

### CODE REVIEW

File: eng/scripts/TypeSpec-Requirement.ps1 [hunk 1]
```diff
@@ -98,27 +98,51 @@ else {
...
+        if ($file -match "specification/(?<specFamily>[^/]+)/") {
+          $specFamily = $Matches["specFamily"];
+          $tspConfigs = @(Get-ChildItem -Path (Join-Path $repoPath "specification" $specFamily) -Recurse -File
+            | Where-Object { $_.Name -eq "tspconfig.yaml" })
+
+          if ($tspConfigs) {
+            LogInfo "  Folder 'specification/$specFamily' contains $($tspConfigs.Count) file(s) named 'tspconfig.yaml'"
+          }
+          else {
+            LogError ("OpenAPI was generated from TypeSpec, but folder 'specification/$specFamily' contains no files name 'tspconfig.yaml'." `
```
Comment (konrad-jamrozik) [inline eng/scripts/TypeSpec-Requirement.ps1:120]:
```suggestion
            LogError ("OpenAPI was generated from TypeSpec, but folder 'specification/$specFamily' contains no files named 'tspconfig.yaml'." `
```

File: eng/scripts/TypeSpec-Requirement.ps1 [hunk 2]
```diff
@@ -98,27 +98,51 @@ else {
...
+              + "  The TypeSpec used to generate OpenAPI must be added to this folder.")
```
Comment (konrad-jamrozik) [inline eng/scripts/TypeSpec-Requirement.ps1:121]: I understand you shorten `TypeSpec spec` and `OpenAPI spec` on purpose?

Like, full sentence would be:

> The TypeSpec spec used to generate OpenAPI spec must be added to this folder

File: eng/scripts/TypeSpec-Requirement.ps1 [hunk 3]
```diff
@@ -98,27 +98,51 @@ else {
...
+      LogWarning "  OpenAPI cannot be parsed as JSON, so assuming not generated from TypeSpec"
+      LogWarning "    $_"
+    }
 
+    if ($jsonContent) {
       if ($null -ne ${jsonContent}?["info"]?["x-typespec-generated"]) {
         LogInfo "  OpenAPI was generated from TypeSpec (contains '/info/x-typespec-generated')"
+
+        if ($file -match "specification/(?<specFamily>[^/]+)/") {
+          $specFamily = $Matches["specFamily"];
```
Comment (konrad-jamrozik) [inline eng/scripts/TypeSpec-Requirement.ps1:112]: I believe the official doc uses the term `RP` / `Resource Provider` instead of `spec family`:

https://github.com/Azure/azure-rest-api-specs/blob/main/README.md#directory-structure

Are we introducing here a new term, the `spec family`, that says something like:

- A "spec family" is usually a "resource provider" as outlined in the official directory structure, but some children dirs of "specification" violate this constraint and add more folders going beyond the "resource provider" requirements. Such cases ale considered spec families.

In other words, all children of specification are roots of spec families, but not all spec families are resource providers.

Perhaps we want to "uplift" the term of "spec family" to more official doc? So we have clear communication going forward on what is what. I wonder if there maybe exist another term for this. Maybe there is just the Resource Provider, but it has strong constraints.

Comment (konrad-jamrozik) [reply eng/scripts/TypeSpec-Requirement.ps1:112]: Actually, there is `service group` concept:

https://github.com/Azure/azure-rest-api-specs/blob/main/README.md#folder-structure-for-service-group

I do wonder how it relates to `spec family`. Is it the same thing (in such case probably we want to use `service group` or are there spec families that aren't even service groups?

File: eng/scripts/TypeSpec-Requirement.ps1 [hunk 4]
```diff
@@ -98,27 +98,51 @@ else {
 
     try {
```
Comment (konrad-jamrozik) [inline eng/scripts/TypeSpec-Requirement.ps1:99]: I think it would be great if you could add somewhere in the header comment of this script that it assumes the parsed directory structure adheres to:

- https://github.com/Azure/azure-rest-api-specs/blob/main/README.md#directory-structure

with some possible differences / violations of rules.

I believe this would help the reader a lot to figure out what this code is trying to achieve.

File: eng/scripts/TypeSpec-Requirement.ps1 (hunk 3 above)
Comment (mikeharder) [reply eng/scripts/TypeSpec-Requirement.ps1:112]: I don't know where I first heard the term "spec family".  I may have invented it.  Pretending I never suggested this name, what name would you recomment?

**Review by konrad-jamrozik — APPROVED** [review]

File: eng/scripts/TypeSpec-Requirement.ps1 (hunk 2 above)
Comment (mikeharder) [reply eng/scripts/TypeSpec-Requirement.ps1:121]: What error message would you use instead?  Consider that GitHub truncates after the words "used to generate" on my screen (see screenshot in PR description).

File: eng/scripts/TypeSpec-Requirement.ps1 (hunk 3 above)
Comment (konrad-jamrozik) [reply eng/scripts/TypeSpec-Requirement.ps1:112]: Maybe `resource provider group` ? Because usually it is `resource provider`. Just an idea.

File: eng/scripts/TypeSpec-Requirement.ps1 (hunk 4 above)
Comment (mikeharder) [reply eng/scripts/TypeSpec-Requirement.ps1:99]: Example file paths are documented here:

https://github.com/Azure/azure-rest-api-specs/pull/28444/files#diff-3c37fefe028ad9bd154b76ea3371f8db2d7a7609dafa323e92d8461880e808b2R82-R84

Do you just want to me to add a link to `README.md#directory-structure` above these examples?

File: eng/scripts/TypeSpec-Requirement.ps1 (hunk 3 above)
Comment (mikeharder) [reply eng/scripts/TypeSpec-Requirement.ps1:112]: I just want a name for the first path segment under `specification`.  If one exists, I want to reuse it.  If not, we should pick a name. "spec family" can be a candidate, but I'd like to solicit more ideas before deciding.  And yes we can change this in a future PR.

File: eng/scripts/TypeSpec-Requirement.ps1 (hunk 2 above)
Comment (konrad-jamrozik) [reply eng/scripts/TypeSpec-Requirement.ps1:121]: I am just saying that I would replace `OpenAPI` with `OpenAPI spec` and similar with `TypeSpec`.

File: eng/scripts/TypeSpec-Requirement.ps1 (hunk 4 above)
Comment (konrad-jamrozik) [reply eng/scripts/TypeSpec-Requirement.ps1:99]: Yeah, that would be great. But I think we probably want to clarify this is "aspirational"; that many exceptions to the directory structure exist as of today in the repo, and this script also tried to handle some of them.

File: eng/scripts/TypeSpec-Requirement.ps1 (hunk 3 above)
Comment (mikeharder) [reply eng/scripts/TypeSpec-Requirement.ps1]: Changed name from "spec family" to "rp folder" for now.

**Comment by konrad-jamrozik** [comment]
Related PR:
- https://github.com/Azure/azure-rest-api-specs/pull/28464
